File: BusinessLogic/Sites/clsHPFFA.vb
Imports System
Imports System.Text
Imports System.Net
Imports System.Xml
Imports System.Data
Imports HtmlAgilityPack
Imports System.Web.HttpUtility

Public Class HPFFA
    Inherits clsFanfic

    Private info As clsFanfic.Story
    Private Browser As clsWeb

#Region "Chapter Navigation"

    Public Overrides Function ProcessChapters(link As String, index As Integer) As String

        Dim htmldoc As String

        link += "&"
        link += "chapter=" & (index + 1)

        htmldoc = GrabData(link)

        Return htmldoc

    End Function

    Public Overrides Function GetChapters(ByVal htmlDoc As String) As String()

        Dim chapters As String()

        Dim doc As HtmlDocument = Nothing
        Dim tdoc As HtmlDocument = Nothing
        Dim nodes As HtmlNodeCollection

        Dim story_url As String

        doc = CleanHTML(htmlDoc)

        nodes = FindNodesByAttribute(doc.DocumentNode, "div", "id", "pagetitle", False)

        tdoc = CleanHTML(nodes(0).InnerHtml)

        nodes = FindLinksByHref(tdoc.DocumentNode, "viewstory.php")

        story_url = nodes(0).Attributes("href").Value

        nodes = FindLinksByHref(doc.DocumentNode, story_url & "&amp;chapter=")

        ReDim chapters(nodes.Count - 1)

        For node_idx = 0 To (nodes.Count - 1)
            chapters(node_idx) = HtmlDecode(nodes(node_idx).Attributes("href").Value)
        Next

        GetStoryInfo(htmlDoc)

        Return chapters

    End Function

    Private Sub GetStoryInfo(htmlDoc As String)

        Dim html As String
        Dim doc As HtmlDocument = Nothing
        Dim tdoc As HtmlDocument = Nothing

        Dim temp As HtmlNodeCollection

        Dim dummy As String
        Dim summary() As String

        Dim ch_idx As Integer
        Dim p_idx As Integer
        Dim u_idx As Integer

        Dim idx As Integer

        doc = CleanHTML(htmlDoc)

        temp = FindNodesByAttribute(doc.DocumentNode, "div", "id", "pagetitle", False)
        tdoc = CleanHTML(temp(0).InnerHtml)
        temp = tdoc.DocumentNode.SelectNodes("//a")

        info.Category = "Harry Potter"

        info.StoryURL = "http://www." & Me.HostName & "/stories/" & temp(0).Attributes("href").Value
        info.AuthorURL = "http://www." & Me.HostName & "/stories/" & temp(1).Attributes("href").Value

        info.Title = temp(0).InnerText
        info.Author = temp(1).InnerText

        temp = FindNodesByAttribute(doc.DocumentNode, "div", "class", "content", False)

        html = temp(2).InnerHtml

        doc = CleanHTML(Html)

        temp = FindNodesByAttribute(doc.DocumentNode, "span", "class", "label", False)

        For temp_idx = 0 To temp.Count - 1

            Html = Replace(Html, temp(temp_idx).OuterHtml, "|" & temp(temp_idx).InnerText)

        Next

        summary = Split(Html, "|")

        For idx = 0 To UBound(summary)
            If InStr(summary(idx), "Chapters:") <> 0 Then
                ch_idx = idx
            End If
            If InStr(summary(idx), "Published:") <> 0 Then
                p_idx = idx
            End If
            If InStr(summary(idx), "Updated:") <> 0 Then
                u_idx = idx
            End If
        Next


        tdoc = CleanHTML(summary(ch_idx))

        dummy = tdoc.DocumentNode.InnerText
        dummy = Mid(dummy, InStr(dummy, ":") + 1)

        info.ChapterCount = Trim(dummy)

        tdoc = CleanHTML(summary(p_idx))
        dummy = tdoc.DocumentNode.InnerText
        dummy = Mid(dummy, InStr(dummy, ">") + 1)
        dummy = Mid(dummy, 1, InStr(dummy, "<") - 1)

        info.PublishDate = CDate(dummy).ToShortDateString()

        tdoc = CleanHTML(summary(u_idx))
        dummy = tdoc.DocumentNode.InnerText
        dummy = Mid(dummy, InStr(dummy, ">") + 1)
        dummy = Mid(dummy, 1, InStr(dummy, "<") - 1)

        info.UpdateDate = CDate(dummy).ToShortDateString()

        info.ID = GetStoryID(info.StoryURL)

    End Sub

#End Region

#Region "Data Extraction"

    Public Overrides Function GrabTitle(htmlDoc As String) As String

        Dim doc As HtmlDocument = Nothing
        Dim nodes As HtmlNodeCollection

        Dim ret As String = ""

        If InStr(htmlDoc, "<select class=""textbox"" name=""chapter""") = 0 Then
            ret = info.Title
        Else

            doc = CleanHTML(htmlDoc)

            nodes = FindNodesByAttribute(doc.DocumentNode, "select", "name", "chapter", False)

            doc = CleanHTML(nodes(0).InnerHtml)

            nodes = FindNodesByAttribute(doc.DocumentNode, "option", "selected")

            If nodes.Count > 0 Then
                ret = info.Title
                ret += "<br /><br />"
                ret += nodes(0).NextSibling.InnerText
            Else
                ret = info.Title
            End If

        End If

        Return ret

    End Function

    Public Overrides Function GrabSeries(htmlDoc As String) As String
        Return info.Category
    End Function

    Public Overrides Function GrabDate(htmlDoc As String, title As String) As String

        Dim ret As String = ""

        Select Case title
            Case "Published: "
                ret = info.PublishDate
            Case "Updated: "
                ret = info.UpdateDate
        End Select

        Return ret

    End Function

    Public Overrides Function GrabAuthor(htmlDoc As String) As String
        Return info.Author
    End Function

    Public Overrides Function GrabBody(htmlDoc As String) As String

        Dim doc As HtmlDocument = Nothing
        Dim nodes As HtmlNodeCollection
        Dim ret As String = ""

        doc = CleanHTML(htmlDoc)

        nodes = FindNodesByAttribute(doc.DocumentNode, "div", "id", "story")

        ret = nodes(0).InnerHtml

        Return ret

    End Function

    Public Overrides Function GetAuthorURL(link As String) As String
        Return info.AuthorURL
    End Function

    Public Overrides Function GetStoryID(link As String) As String

        Dim ret As String = ""
        Dim url As URL

        url = ExtractUrl(link)
        ret = url.Query(0).Value

        Return ret

    End Function

    Public Overrides Function GetStoryURL(id As String) As String

        Return "http://www." & Me.HostName & "/stories/viewstory.php?sid=" & id

    End Function


#End Region

#Region "Download Routines"

    Protected Overrides Function GrabFeedData(ByRef rss As String) As System.Xml.XmlDocument

        Dim html As String
        Dim doc As HtmlDocument = Nothing
        Dim tdoc As HtmlDocument = Nothing

        Dim nodes As HtmlNodeCollection
        Dim temp As HtmlNodeCollection

        Dim dummy As String

        Dim temp_idx As Integer
        Dim node_idx As Integer

        Dim author_url As String

        Dim summary() As String

        Dim ch_idx As Integer

        Dim fic() As clsFanfic.Story

        Dim idx As Integer

        Dim xmldoc As XmlDocument = Nothing

        author_url = rss

        html = Me.GrabData(rss)

        doc = CleanHTML(html)

        nodes = FindNodesByAttribute(doc.DocumentNode, "div", "class", "listbox", True)

        ReDim fic(nodes.Count - 1)

        For node_idx = 0 To nodes.Count - 1

            doc = CleanHTML(nodes(node_idx).InnerHtml)

            temp = FindNodesByAttribute(doc.DocumentNode, "div", "class", "title", False)
            tdoc = CleanHTML(temp(0).InnerHtml)
            temp = tdoc.DocumentNode.SelectNodes("//a")

            fic(node_idx).Category = "Harry Potter"

            fic(node_idx).StoryURL = "http://www." & Me.HostName & "/stories/" & temp(0).Attributes("href").Value
            fic(node_idx).AuthorURL = "http://www." & Me.HostName & "/stories/" & temp(1).Attributes("href").Value

            fic(node_idx).Title = temp(0).InnerText
            fic(node_idx).Author = temp(1).InnerText


            temp = FindNodesByAttribute(doc.DocumentNode, "div", "class", "tail", False)

            dummy = temp(0).InnerHtml

            tdoc = CleanHTML(dummy)

            temp = FindNodesByAttribute(doc.DocumentNode, "span", "class", "label", False)

            For temp_idx = 0 To temp.Count - 1

                dummy = Replace(dummy, temp(temp_idx).OuterHtml, "|" & temp(temp_idx).InnerText)

            Next

            dummy = Replace(dummy, "</span>", "")
            summary = Split(dummy, "|")

            fic(node_idx).PublishDate = CDate(Mid(summary(1), InStr(summary(1), ":") + 1)).ToShortDateString
            fic(node_idx).UpdateDate = CDate(Mid(summary(2), InStr(summary(2), ":") + 1)).ToShortDateString

            temp = FindNodesByAttribute(doc.DocumentNode, "div", "class", "content", False)

            html = temp(0).InnerHtml

            doc = CleanHTML(html)

            temp = FindNodesByAttribute(doc.DocumentNode, "span", "class", "label", False)

            For temp_idx = 0 To temp.Count - 1

                html = Replace(html, temp(temp_idx).OuterHtml, "|" & temp(temp_idx).InnerText)

            Next

            summary = Split(html, "|")

            tdoc = CleanHTML(summary(1))

            dummy = tdoc.DocumentNode.InnerText & vbCrLf

            For idx = 0 To UBound(summary)
                If InStr(summary(idx), "Chapters:") <> 0 Then
                    ch_idx = idx
                    Exit For
                End If
            Next

            For idx = 2 To (ch_idx - 2)

                If idx <> 4 Then

                    tdoc = CleanHTML(summary(idx))
                    dummy += HtmlDecode(tdoc.DocumentNode.InnerText) & vbCrLf

                End If

            Next

            fic(node_idx).Summary = dummy

            tdoc = CleanHTML(summary(ch_idx))

            dummy = tdoc.DocumentNode.InnerText
            dummy = Left(dummy, InStr(dummy, "Table") - 1)
            dummy = Mid(dummy, InStr(dummy, ":") + 1)

            fic(node_idx).ChapterCount = Trim(dummy)

            fic(node_idx).ID = GetStoryID(fic(node_idx).StoryURL)

        Next node_idx

        html = GenerateAtomFeed(fic)

        doc = CleanHTML(html)

        html = doc.DocumentNode.OuterHtml

        xmldoc = New XmlDocument

        xmldoc.LoadXml(html)

        Return xmldoc

    End Function

#End Region


#Region "RSS Routines"

    Public Overrides Function GrabStoryInfo(idx As Integer) As clsFanfic.Story

        Dim dsRSS As DataSet = MyBase.datasetRSS

        Dim fic As New clsFanfic.Story

        Dim temp() As String

        ' Story Name
        fic.Title = dsRSS.Tables("entry"). _
                        Rows(idx).Item("title")

        ' Story Author
        fic.Author = dsRSS.Tables("author"). _
                          Rows(idx).Item(0)
        'Story Author URL
        fic.AuthorURL = dsRSS.Tables("author"). _
                        Rows(idx).Item(1)

        ' Story Location
        fic.StoryURL = dsRSS.Tables("link"). _
                          Rows(idx).Item(1)

        fic.ID = Me.GetStoryID(fic.StoryURL)

        fic.PublishDate = CDate(dsRSS.Tables("entry").Rows(idx).Item("published"))

        fic.UpdateDate = CDate(dsRSS.Tables("entry").Rows(idx).Item("updated"))

        temp = Split(dsRSS.Tables("entry").Rows(idx).Item("id"), ":")

        fic.Category = temp(1)
        fic.ChapterCount = temp(2)

        fic.Summary = dsRSS.Tables("summary").Rows(idx).Item("summary_Text")

        Return fic

    End Function

#End Region

#Region "Properties"

    Public Overrides ReadOnly Property HostName() As String
        Get
            Return "hpfanficarchive.com"
        End Get
    End Property

    Public Overrides ReadOnly Property Name() As String
        Get
            Return "HPFFA"
        End Get
    End Property

    Public Overrides ReadOnly Property ErrorMessage As String
        Get
            Return "You are not authorized to access that function."
        End Get
    End Property

#End Region

    Public Sub New()

        Browser = New clsWeb

    End Sub
End Class

File: FFDownload/HtmlScraper/BusinessLogic/Sites/clsHPFFA.cs
using System.Xml;
using HtmlAgilityPack;
using HtmlScraper.BusinessLogic.Sites.Base;
using HtmlScraper.Utility.Browser.HTML;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using static System.Web.HttpUtility;

namespace HtmlScraper.BusinessLogic.Sites
{

    public class HPFFA : clsFanfic
    {

        private Story info;

        #region Chapter Navigation

        public override string ProcessChapters(string link, int index)
        {

            string htmldoc;

            link += "&";
            link += "chapter=" + (index + 1);

            htmldoc = GrabData(link);

            return htmldoc;

        }

        public override string[] GetChapters(string htmlDoc)
        {

            string[] chapters;

            HtmlAgilityPack.HtmlDocument doc = null;
            HtmlAgilityPack.HtmlDocument tdoc = null;
            HtmlNodeCollection nodes;

            string story_url;

            doc = modHTML.CleanHTML(ref htmlDoc);

            nodes = modHTML.FindNodesByAttribute(doc.DocumentNode, "div", "id", "pagetitle", false);

            string arghtml = nodes[0].InnerHtml;
            tdoc = modHTML.CleanHTML(ref arghtml);
            nodes[0].InnerHtml = arghtml;

            nodes = modHTML.FindLinksByHref(tdoc.DocumentNode, "viewstory.php");

            story_url = nodes[0].Attributes["href"].Value;

            nodes = modHTML.FindLinksByHref(doc.DocumentNode, story_url + "&amp;chapter=");

            chapters = new string[nodes.Count];

            for (int node_idx = 0, loopTo = nodes.Count - 1; node_idx <= loopTo; node_idx++)
                chapters[node_idx] = HtmlDecode(nodes[node_idx].Attributes["href"].Value);

            GetStoryInfo(htmlDoc);

            return chapters;

        }

        private void GetStoryInfo(string htmlDoc)
        {

            string html;
            HtmlAgilityPack.HtmlDocument doc = null;
            HtmlAgilityPack.HtmlDocument tdoc = null;

            HtmlNodeCollection temp;

            string dummy;
            string[] summary;

            var ch_idx = default(int);
            var p_idx = default(int);
            var u_idx = default(int);

            int idx;

            doc = modHTML.CleanHTML(ref htmlDoc);

            temp = modHTML.FindNodesByAttribute(doc.DocumentNode, "div", "id", "pagetitle", false);
            string arghtml = temp[0].InnerHtml;
            tdoc = modHTML.CleanHTML(ref arghtml);
            temp[0].InnerHtml = arghtml;
            temp = tdoc.DocumentNode.SelectNodes("//a");

            info.Category = "Harry Potter";

            info.StoryURL = "http://www." + HostName + "/stories/" + temp[0].Attributes["href"].Value;
            info.AuthorURL = "http://www." + HostName + "/stories/" + temp[1].Attributes["href"].Value;

            info.Title = temp[0].InnerText;
            info.Author = temp[1].InnerText;

            temp = modHTML.FindNodesByAttribute(doc.DocumentNode, "div", "class", "content", false);

            html = temp[2].InnerHtml;

            doc = modHTML.CleanHTML(ref html);

            temp = modHTML.FindNodesByAttribute(doc.DocumentNode, "span", "class", "label", false);

            for (int temp_idx = 0, loopTo = temp.Count - 1; temp_idx <= loopTo; temp_idx++)


                html = Strings.Replace(html, temp[temp_idx].OuterHtml, "|" + temp[temp_idx].InnerText);

            summary = Strings.Split(html, "|");

            var loopTo1 = Information.UBound(summary);
            for (idx = 0; idx <= loopTo1; idx++)
            {
                if (Strings.InStr(summary[idx], "Chapters:") != 0)
                {
                    ch_idx = idx;
                }
                if (Strings.InStr(summary[idx], "Published:") != 0)
                {
                    p_idx = idx;
                }
                if (Strings.InStr(summary[idx], "Updated:") != 0)
                {
                    u_idx = idx;
                }
            }


            tdoc = modHTML.CleanHTML(ref summary[ch_idx]);

            dummy = tdoc.DocumentNode.InnerText;
            dummy = Strings.Mid(dummy, Strings.InStr(dummy, ":") + 1);

            info.ChapterCount = Strings.Trim(dummy);

            tdoc = modHTML.CleanHTML(ref summary[p_idx]);
            dummy = tdoc.DocumentNode.InnerText;

            dummy = Strings.Replace(dummy, "Published:", "");

            if (Strings.InStr(dummy, ">") > 0)
            {

                dummy = Strings.Mid(dummy, Strings.InStr(dummy, ">") + 1);
                dummy = Strings.Mid(dummy, 1, Strings.InStr(dummy, "<") - 1);

            }

            info.PublishDate = Conversions.ToDate(dummy).ToShortDateString();

            tdoc = modHTML.CleanHTML(ref summary[u_idx]);
            dummy = tdoc.DocumentNode.InnerText;

            if (Strings.InStr(dummy, ">") > 0)
            {

                dummy = Strings.Mid(dummy, Strings.InStr(dummy, ">") + 1);
                dummy = Strings.Mid(dummy, 1, Strings.InStr(dummy, "<") - 1);

            }

            dummy = Strings.Replace(dummy, "Updated:", "");

            info.UpdateDate = Conversions.ToDate(dummy).ToShortDateString();

            info.ID = GetStoryID(info.StoryURL);

        }

        #endregion

        #region Data Extraction

        public override string GrabTitle(string htmlDoc)
        {

            HtmlAgilityPack.HtmlDocument doc = null;
            HtmlNodeCollection nodes;

            string ret = "";

            if (Strings.InStr(htmlDoc, "<select class=\"textbox\" name=\"chapter\"") == 0)
            {
                ret = info.Title;
            }
            else
            {

                doc = modHTML.CleanHTML(ref htmlDoc);

                nodes = modHTML.FindNodesByAttribute(doc.DocumentNode, "select", "name", "chapter", false);

                string arghtml = nodes[0].InnerHtml;
                doc = modHTML.CleanHTML(ref arghtml);
                nodes[0].InnerHtml = arghtml;

                nodes = modHTML.FindNodesByAttribute(doc.DocumentNode, "option", "selected");

                if (!(nodes == null))
                {
                    if (nodes.Count > 0)
                    {
                        ret = info.Title;
                        ret += "<br /><br />";
                        ret += nodes[0].NextSibling.InnerText;
                    }
                    else
                    {
                        ret = info.Title;
                    }
                }
                else
                {
                    ret = info.Title;
                }

                return ret;

            }

            return ret;

        }

        public override string GrabSeries(string htmlDoc)
        {
            return info.Category;
        }

        public override string GrabDate(string htmlDoc, string title)
        {

            string ret = "";

            switch (title ?? "")
            {
                case "Published: ":
                    {
                        ret = info.PublishDate;
                        break;
                    }
                case "Updated: ":
                    {
                        ret = info.UpdateDate;
                        break;
                    }
            }

            return ret;

        }

        public override string GrabAuthor(string htmlDoc)
        {
            return info.Author;
        }

        public override string GrabBody(string htmlDoc)
        {

            HtmlAgilityPack.HtmlDocument doc = null;
            HtmlNodeCollection nodes;
            string ret = "";

            doc = modHTML.CleanHTML(ref htmlDoc);

            nodes = modHTML.FindNodesByAttribute(doc.DocumentNode, "div", "id", "story");

            ret = nodes[0].InnerHtml;

            return ret;

        }

        public override string GetAuthorURL(string link)
        {
            return info.AuthorURL;
        }

        public override string GetStoryID(string link)
        {

            string ret = "";
            URL url;

            url = URLHelper.ExtractUrl(link);
            ret = url.Query[0].Value;

            return ret;

        }

        public override string GetStoryURL(string id)
        {

            return "http://www." + HostName + "/stories/viewstory.php?sid=" + id;

        }


        #endregion

        #region Download Routines

        protected override XmlDocument GrabFeedData(ref string rss)
        {

            string html;
            HtmlAgilityPack.HtmlDocument doc = null;
            HtmlAgilityPack.HtmlDocument tdoc = null;

            HtmlNodeCollection nodes;
            HtmlNodeCollection temp;

            string dummy;

            int temp_idx;
            int node_idx;

            string author_url;

            string[] summary;

            var ch_idx = default(int);

            Story[] fic;

            int idx;

            XmlDocument xmldoc = null;

            author_url = rss;

            if (Strings.InStr(rss, "action=storiesby") == 0 & Strings.InStr(rss, "viewuser.php") > 0)
            {
                rss += "&action=storiesby";
            }
            html = GrabData(rss);

            doc = modHTML.CleanHTML(ref html);

            nodes = modHTML.FindNodesByAttribute(doc.DocumentNode, "div", "class", "listbox", true);

            fic = new Story[nodes.Count];

            var loopTo = nodes.Count - 1;
            for (node_idx = 0; node_idx <= loopTo; node_idx++)
            {

                string arghtml = nodes[node_idx].InnerHtml;
                doc = modHTML.CleanHTML(ref arghtml);
                nodes[node_idx].InnerHtml = arghtml;

                temp = modHTML.FindNodesByAttribute(doc.DocumentNode, "div", "class", "title", false);
                string arghtml1 = temp[0].InnerHtml;
                tdoc = modHTML.CleanHTML(ref arghtml1);
                temp[0].InnerHtml = arghtml1;
                temp = tdoc.DocumentNode.SelectNodes("//a");

                fic[node_idx].Category = "Harry Potter";

                fic[node_idx].StoryURL = "http://www." + HostName + "/stories/" + temp[0].Attributes["href"].Value;
                fic[node_idx].AuthorURL = "http://www." + HostName + "/stories/" + temp[1].Attributes["href"].Value;

                fic[node_idx].Title = temp[0].InnerText;
                fic[node_idx].Author = temp[1].InnerText;


                temp = modHTML.FindNodesByAttribute(doc.DocumentNode, "div", "class", "tail", false);

                dummy = temp[0].InnerHtml;

                tdoc = modHTML.CleanHTML(ref dummy);

                temp = modHTML.FindNodesByAttribute(doc.DocumentNode, "span", "class", "label", false);

                var loopTo1 = temp.Count - 1;
                for (temp_idx = 0; temp_idx <= loopTo1; temp_idx++)


                    dummy = Strings.Replace(dummy, temp[temp_idx].OuterHtml, "|" + temp[temp_idx].InnerText);

                dummy = Strings.Replace(dummy, "</span>", "");
                summary = Strings.Split(dummy, "|");

                fic[node_idx].PublishDate = Conversions.ToDate(Strings.Mid(summary[1], Strings.InStr(summary[1], ":") + 1)).ToShortDateString();
                fic[node_idx].UpdateDate = Conversions.ToDate(Strings.Mid(summary[2], Strings.InStr(summary[2], ":") + 1)).ToShortDateString();

                temp = modHTML.FindNodesByAttribute(doc.DocumentNode, "div", "class", "content", false);

                html = temp[0].InnerHtml;

                doc = modHTML.CleanHTML(ref html);

                temp = modHTML.FindNodesByAttribute(doc.DocumentNode, "span", "class", "label", false);

                var loopTo2 = temp.Count - 1;
                for (temp_idx = 0; temp_idx <= loopTo2; temp_idx++)


                    html = Strings.Replace(html, temp[temp_idx].OuterHtml, "|" + temp[temp_idx].InnerText);

                summary = Strings.Split(html, "|");

                tdoc = modHTML.CleanHTML(ref summary[1]);

                dummy = tdoc.DocumentNode.InnerText + Constants.vbCrLf;

                var loopTo3 = Information.UBound(summary);
                for (idx = 0; idx <= loopTo3; idx++)
                {
                    if (Strings.InStr(summary[idx], "Chapters:") != 0)
                    {
                        ch_idx = idx;
                        break;
                    }
                }

                var loopTo4 = ch_idx - 2;
                for (idx = 2; idx <= loopTo4; idx++)
                {

                    if (idx != 4)
                    {

                        tdoc = modHTML.CleanHTML(ref summary[idx]);
                        dummy += HtmlDecode(tdoc.DocumentNode.InnerText) + Constants.vbCrLf;

                    }

                }

                fic[node_idx].Summary = dummy;

                tdoc = modHTML.CleanHTML(ref summary[ch_idx]);

                dummy = tdoc.DocumentNode.InnerText;
                dummy = Strings.Left(dummy, Strings.InStr(dummy, "Table") - 1);
                dummy = Strings.Mid(dummy, Strings.InStr(dummy, ":") + 1);

                fic[node_idx].ChapterCount = Strings.Trim(dummy);

                fic[node_idx].ID = GetStoryID(fic[node_idx].StoryURL);

            }

            html = GenerateAtomFeed(fic);

            doc = modHTML.CleanHTML(ref html);

            html = doc.DocumentNode.OuterHtml;

            xmldoc = new XmlDocument();

            xmldoc.LoadXml(html);

            return xmldoc;

        }

        #endregion


        #region RSS Routines

        public override Story GrabStoryInfo(int idx)
        {

            var dsRSS = datasetRSS;

            var fic = new Story();

            string[] temp;

            // Story Name
            fic.Title = Conversions.ToString(dsRSS.Tables["entry"].Rows[idx]["title"]);

            // Story Author
            fic.Author = Conversions.ToString(dsRSS.Tables["author"].Rows[idx][0]);
            // Story Author URL
            fic.AuthorURL = Conversions.ToString(dsRSS.Tables["author"].Rows[idx][1]);

            // Story Location
            fic.StoryURL = Conversions.ToString(dsRSS.Tables["link"].Rows[idx][1]);

            fic.ID = GetStoryID(fic.StoryURL);

            fic.PublishDate = Conversions.ToString(Conversions.ToDate(dsRSS.Tables["entry"].Rows[idx]["published"]));

            fic.UpdateDate = Conversions.ToString(Conversions.ToDate(dsRSS.Tables["entry"].Rows[idx]["updated"]));

            temp = Strings.Split(Conversions.ToString(dsRSS.Tables["entry"].Rows[idx]["id"]), ":");

            fic.Category = temp[1];
            fic.ChapterCount = temp[2];

            fic.Summary = Conversions.ToString(dsRSS.Tables["summary"].Rows[idx]["summary_Text"]);

            return fic;

        }

        #endregion

        #region Properties

        public override string HostName
        {
            get
            {
                return "hpfanficarchive.com";
            }
        }

        public override string Name
        {
            get
            {
                return "HPFFA";
            }
        }

        public override string ErrorMessage
        {
            get
            {
                return "does not exist on this archive";
            }
        }

        #endregion

    }
}